File: csharp/Dson/src/DsonExtDouble.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using Wjybxx.Dson.Types;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson具有类型标签的双精度浮点数
/// </summary>
public sealed class DsonExtDouble : DsonValue, IEquatable<DsonExtDouble>, IComparable<DsonExtDouble>, IComparable
{
    private readonly ExtDouble _extDouble;

    public DsonExtDouble(int type, double? value)
        : this(new ExtDouble(type, value)) {
    }

    public DsonExtDouble(int type, double value, bool hasVal = true)
        : this(new ExtDouble(type, value, hasVal)) {
    }

    public DsonExtDouble(ExtDouble extDouble) {
        _extDouble = extDouble;
    }

    public ExtDouble ExtDouble => _extDouble;
    public override DsonType DsonType => DsonType.ExtDouble;
    public int Type => _extDouble.Type;
    public bool HasValue => _extDouble.HasValue;
    public double Value => _extDouble.Value;

    #region equals

    public bool Equals(DsonExtDouble? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _extDouble.Equals(other._extDouble);
    }

    public override bool Equals(object? obj) {
        return ReferenceEquals(this, obj) || obj is DsonExtDouble other && Equals(other);
    }

    public override int GetHashCode() {
        return _extDouble.GetHashCode();
    }

    public static bool operator ==(DsonExtDouble? left, DsonExtDouble? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonExtDouble? left, DsonExtDouble? right) {
        return !Equals(left, right);
    }

    public int CompareTo(DsonExtDouble? other) {
        if (ReferenceEquals(this, other)) return 0;
        if (ReferenceEquals(null, other)) return 1;
        return _extDouble.CompareTo(other._extDouble);
    }

    public int CompareTo(object? obj) {
        if (ReferenceEquals(null, obj)) return 1;
        if (ReferenceEquals(this, obj)) return 0;
        return obj is DsonExtDouble other ? CompareTo(other) : throw new ArgumentException($"Object must be of type {nameof(DsonExtDouble)}");
    }

    #endregion

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_extDouble)}: {_extDouble}";
    }
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonExtDouble.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.types.ExtDouble;

import javax.annotation.Nonnull;
import java.util.Objects;

/**
 * double的简单扩展
 * double虽然支持NaN，但和hasValue的含义还是不一样。
 *
 * @author wjybxx
 * date - 2023/12/2
 */
public final class DsonExtDouble extends DsonValue implements Comparable<DsonExtDouble> {

    private final ExtDouble extDouble;

    public DsonExtDouble(int type, double value) {
        this(new ExtDouble(type, value));
    }

    public DsonExtDouble(int type, Double value) {
        this(new ExtDouble(type, value));
    }

    public DsonExtDouble(int type, double value, boolean hasValue) {
        this(new ExtDouble(type, value, hasValue));
    }

    public DsonExtDouble(ExtDouble extDouble) {
        this.extDouble = Objects.requireNonNull(extDouble);
    }

    public ExtDouble extDouble() {
        return extDouble;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.EXT_DOUBLE;
    }

    public int getType() {
        return extDouble.getType();
    }

    public double getValue() {
        return extDouble.getValue();
    }

    public boolean hasValue() {
        return extDouble.hasValue();
    }

    //region equals

    @Override
    public int compareTo(DsonExtDouble that) {
        return extDouble.compareTo(that.extDouble);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonExtDouble that = (DsonExtDouble) o;

        return extDouble.equals(that.extDouble);
    }

    @Override
    public int hashCode() {
        return extDouble.hashCode();
    }

    // endregion

    @Override
    public String toString() {
        return "DsonExtDouble{" +
                "extDouble=" + extDouble +
                '}';
    }
}
